crypto/crypto.c

```diff
@@ -0,0 +1,425 @@
+{
+    int rc;
+    BIO *bio = BIO_new(BIO_s_mem());
+    char *tmp = NULL;
+    rc = X509_print_ex(bio,(X509 *)x509, XN_FLAG_MULTILINE, X509_FLAG_COMPAT | X509_FLAG_NO_PUBKEY | X509_FLAG_NO_SIGDUMP);
+    if (rc < 0){
+        prlog(PR_ERR, "ERROR: could not get BIO data on X509, openssl err#%d\n",rc);
+        return rc;
+    }
+    //returns total data avialable
+    rc = BIO_get_mem_data(bio, &tmp);
+    memcpy(x509_info, tmp, max_len > rc ? rc : max_len -1);
+    BIO_free(bio);
+    return max_len > rc ? rc : max_len -1;
+}
+
+void *crypto_x509_parse_der(const unsigned char *data, size_t data_len)
+{
+    X509* x509;
+    x509 = d2i_X509(NULL, &data, data_len);
+    
+    if (!x509)
+        return NULL;
+
+    return (void *)x509; 
+}
+
+void crypto_x509_free(void *x509)
+{
+    X509_free((X509 *)x509);
+}
+
+int crypto_convert_pem_to_der(const unsigned char *input, size_t ilen, unsigned char **output, size_t *olen)
+{
+    int rc;
+    BIO *bio;
+    bio = BIO_new_mem_buf(input, ilen);
+    //these variables are not needed on return, just needed to properly call the function
+    char *header = NULL, *name = NULL;
+    //returns 0 for fail and 1 on success
+    rc = !PEM_read_bio(bio, &name, &header, output, (long int *)olen);
+    if (header) free(header);
+    if (name) free(name);
+    BIO_free(bio);
+    return rc;
+}
+
+/*=====================END OPENSSL FUNCTIONS=====================*/
+#else
+/*====================START MBEDTLS FUNCTIONS====================*/
+
+
+
+void *crypto_pkcs7_parse_der(const unsigned char *buf, const int buflen) 
+{
+    int rc;
+    struct mbedtls_pkcs7 *pkcs7;
+    pkcs7 = malloc(sizeof(struct mbedtls_pkcs7));
+    if (!pkcs7) {
+     prlog(PR_ERR, "ERROR: failed to allocate memory\n");
+     return NULL;
+    }
+    mbedtls_pkcs7_init(pkcs7);
+    rc = mbedtls_pkcs7_parse_der(buf, buflen, pkcs7);
+    if (rc != MBEDTLS_PKCS7_SIGNED_DATA)  // if pkcs7 parsing fails, then try new signed data format 
+            prlog(PR_ERR, "ERROR: parsing pkcs7 failed mbedtls error #%04x\n", rc); 
+    else 
+        rc = SUCCESS;
+
+    if (rc) {
+        crypto_pkcs7_free(pkcs7);
+        return NULL;
+    }
+    else
+        return (void *)pkcs7;
+}
+
+// size_t crypto_get_pkcs7_struct_size() 
+// {
+//     return sizeof(struct mbedtls_pkcs7);
+// }
+
+int crypto_pkcs7_is_md_sha256(void *pkcs7)
+{
+    return memcmp(((struct mbedtls_pkcs7 *)pkcs7)->signed_data.digest_alg_identifiers.p, MBEDTLS_OID_DIGEST_ALG_SHA256, strlen(MBEDTLS_OID_DIGEST_ALG_SHA256) ) ;
+}
+void crypto_pkcs7_free(void *pkcs7)
+{
+        mbedtls_pkcs7_free((struct mbedtls_pkcs7 *)pkcs7);
+        free(pkcs7);
+}
+
+void* crypto_get_signing_cert(void *pkcs7, int cert_num)
+{
+    mbedtls_x509_crt *pkcs7_cert = NULL;
+
+    pkcs7_cert = &((struct mbedtls_pkcs7 *)pkcs7)->signed_data.certs;
+    for (int i = 0; i < cert_num && pkcs7_cert != NULL; i++)
+        pkcs7_cert = pkcs7_cert->next;
+
```

Comment:
What should happen if you ask for cert number 42 in a pkcs7 with only 3 certs? you'll get cert 3 atm, which may or may not be what you want? not sure what openssl does here, they should at least be consistent.

Reply:
In this function, if the requested index is greater than the actual, NULL would be returned. So in your example, NULL is returned not the third certificate since the loop exits if the current cert is NULL or if we are at the desired index. Openssl also returns NULL if the desired index is out of range: `sk_TYPE_value() returns element idx in sk, where idx starts at zero. If idx is out of range then NULL is returned`